app/src/main/java/tech/relaycorp/letro/awala/AwalaManager.kt

```diff
@@ -0,0 +1,228 @@
+package tech.relaycorp.letro.awala
+
+import android.content.Context
+import android.util.Log
+import androidx.annotation.RawRes
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+import tech.relaycorp.awaladroid.Awala
+import tech.relaycorp.awaladroid.GatewayBindingException
+import tech.relaycorp.awaladroid.GatewayClient
+import tech.relaycorp.awaladroid.endpoint.FirstPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.InvalidThirdPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.PublicThirdPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.ThirdPartyEndpoint
+import tech.relaycorp.awaladroid.messaging.OutgoingMessage
+import tech.relaycorp.letro.R
+import tech.relaycorp.letro.awala.message.Message
+import tech.relaycorp.letro.awala.message.MessageRecipient
+import tech.relaycorp.letro.awala.message.MessageType
+import tech.relaycorp.letro.ui.navigation.Route
+import tech.relaycorp.letro.utils.awala.loadNonNullFirstPartyEndpoint
+import tech.relaycorp.letro.utils.awala.loadNonNullThirdPartyEndpoint
+import javax.inject.Inject
+
+interface AwalaManager {
+    val messages: Flow<Message>
+    suspend fun sendMessage(
+        message: Message,
+        recipient: MessageRecipient,
+    )
+    suspend fun isAwalaInstalled(currentScreen: Route): Boolean
+}
+
+class AwalaManagerImpl @Inject constructor(
+    private val awalaRepository: AwalaRepository,
+    @ApplicationContext private val context: Context,
+): AwalaManager {
+
+    private val awalaScope = CoroutineScope(Dispatchers.IO)
+
+    private val _messages = MutableSharedFlow<Message>()
+    override val messages: Flow<Message>
+        get() = _messages
+
+    @Volatile
+    private var isAwalaSetUp = false
+    private var awalaSetupJob: Job? = null
+
+    @Volatile
+    private var isAwalaInstalledOnDevice: Boolean? = null
+
+    @Volatile
+    private var isReceivingMessages = false
+
+    init {
+        awalaSetupJob = awalaScope.launch {
+            Awala.setUp(context)
+            checkIfAwalaAppInstalled()
+            isAwalaSetUp = true
+            awalaSetupJob = null
+        }
+    }
+
+    override suspend fun sendMessage(
+        message: Message,
+        recipient: MessageRecipient,
+    ) {
+        val firstPartyEndpointNodeId = awalaRepository.getServerFirstPartyEndpointNodeId()
+            ?: registerFirstPartyEndpointIfNeeded()?.nodeId
+            ?: ""
```

Comment:
How about we just do:

```suggestion
            ?: throw IllegalStateException("First-party endpoint id is missing")
```

And the same with the third party endpoint.

This would simplify the code.

Reply:
Good point!